parlai/core/worlds.py

```diff
@@ -975,6 +975,202 @@ def shutdown(self):
+            # while not really wasting much padding.
+            return self._ceil(len(obs['text_vec']))
+        else:
+            return None
+
+    def parley(self):
+        # first make sure that all the worlds are processed in the queue
+        assert not self.epoch_done()
+        indices = []
+        for i in range(self._BUFFER_SIZE):
+            if self._scores[i] is not None:
+                indices.append(i)
+                continue
+            if self.worlds[i].epoch_done():
+                continue
+
+            act = self.worlds[i].agents[0].act()
+            obs = self.worlds[i].agents[1].observe(act)
+            self._obs[i] = obs
+
+            if act is None:
+                # TODO: we might need to do something special here around the end of
+                # the epoch
+                raise TypeError("This is unexpected.")
+
+            self._scores[i] = self._score(obs)
+            if self._scores[i] is not None:
+                indices.append(i)
+
+        assert len(indices) != 0, "DynamicBatchWorld ran out of data!"
+
+        for i in indices:
+            assert self._scores is not None
+
+        # sort all the indices by their score, so that we can find similarly lengthed
+        # items in O(1)
+        indices = sorted(indices, key=lambda i: (self._scores[i], random.random()))
+        # indices = sorted(indices, key=lambda i: (random.random()))
+
+        # now let's build the batch
+        batch = []
+
+        # start with a random item. index_index is the lookup into indices, but
+        # index is the actual world.
+        width = 0
+        index_index = random.randint(0, len(indices) - 1)
+
+        # we picked a random spot, but we can get better packing if we start at the
+        # last example with the same score, since we always pick move down to smaller examples.
+        while index_index < len(indices) - 1 and (
+            self._scores[indices[index_index]] == self._scores[indices[index_index + 1]]
+        ):
+            index_index += 1
+
+        # quit early if we eat our full buffer
+        while indices:
+            index = indices[index_index]
+            this_width = self._ceil(self._scores[index])
+            new_width = max(width, this_width)
+            # compute the cost of the new batch
+            new_size = new_width * self._ceil(len(batch) + 1)
+            # TODO: delete this:
+            # if len(batch) < self.opt['batchsize']:
+            if new_size <= self.max_size:
+                # cool, this one fits, let's add it
+                width = new_width
+                batch.append(index)
+                indices.pop(index_index)
+                index_index = max(index_index - 1, 0)
+            else:
+                # we'd overfill our buffer, give up
+                break
+
+        # double check our assumed invariant
+        assert self._ceil(width) * self._ceil(len(batch)) <= self.max_size
```

Comment:
is it even possible that this wouldn't be true?

Reply:
no, it's more there for my sanity to annotate the invariant condition